[go_router] Don't log if `hierarchicalLoggingEnabled` is `true`

ValentinVignal opened the pull request:
Fixes https://github.com/flutter/flutter/issues/139667

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

ValentinVignal commented:
I would like to ask for some help with the tests here. 

Here is the `setLogging` method: 

https://github.com/flutter/packages/blob/5b48c446976fd5f61316d58ab3fcc9888423c2e6/packages/go_router/lib/src/logging.dart#L27-L62


My idea was to write 2 tests (both with `_enabled` set to `true`):
- 1 with `hierarchicalLoggingEnabled` set to `false` and verify that `developer.log` is called or something is logged in the console.
- 1 with `hierarchicalLoggingEnabled` set to `true` and verify that `developer.log` is not called or nothing is logged in the console.

I first wrote this test:

```dart
  testWidgets(
    'It should log the known routes and the initial route if debugLogDiagnostics is false',
    (WidgetTester tester) async {
      final _FakeStdout stdout = _FakeStdout();
      IOOverrides.runZoned(
        () {
          final List<String> logs = <String>[];
          Logger.root.onRecord.listen(
            (LogRecord event) => logs.add(event.message),
          );
          GoRouter(
            debugLogDiagnostics: true,
            routes: <RouteBase>[
              GoRoute(
                path: '/',
                builder: (_, GoRouterState state) => const Text('home'),
              ),
            ],
          );

          expect(
            logs,
            const <String>[
              'Full paths for routes:\n  => /\n',
              'setting initial location null'
            ],
          );
          expect(stdout.writtenObjects, isNotEmpty);
        },
        stdout: () => stdout,
      );
    },
  );
```

But `stdout.writtenObjects` is empty. I cannot find a way to verify that `developer.log` is called.

Do you have any suggestion?

ValentinVignal commented:
@chunhtai Any idea how I can write a test for that?

ValentinVignal commented:
Hello @chunhtai, do you have any thoughts on this?

cedvdb commented:
I think you have to get someone on discord for the review

ValentinVignal commented:
@cedvdb Would you know which channel I should use ? I tried [hackers-routing-🛤️ ](https://discord.com/channels/608014603317936148/969301283825659914), but no one answered

hangyujin commented:
@ValentinVignal I will take a look at this PR, also @chunhtai will be back in a week

chunhtai commented on packages/go_router/lib/src/route_data.dart:
```diff
@@ -63,6 +63,11 @@ abstract class GoRouteData extends RouteData {
+  /// Called when this route is removed from GoRouter's route history.
+  ///
+  /// Corresponds to [GoRoute.onExit].
+  FutureOr<bool> onExit(BuildContext context) => true;
```
is this from other patch? we should probably separate this out

ValentinVignal replied:
Oops yeah, that's a mistake. I reverted it

chunhtai commented on packages/go_router/lib/src/delegate.dart:
```diff
@@ -72,6 +72,8 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
     // This should be the only place where the last GoRoute exit the screen.
     final GoRoute lastRoute = currentConfiguration.last.route;
     if (lastRoute.onExit != null && navigatorKey.currentContext != null) {
+      walker.buildState(_configuration, _configuration.);
```
This seems unrelated?

ValentinVignal commented:
@chunhtai I refactored `logging` to be able to mock `developer.log` in [refactor: Add mockable developerLog method](https://github.com/flutter/packages/pull/6019/commits/55e67307c20109863a4d2275421b1c96cff83805) and added some tests in [test: Update the tests](https://github.com/flutter/packages/pull/6019/commits/213841feb44823166d3932f77c015074f0a4c487). Is that what you were expecting?

chunhtai reviewed (approved):
LGTM

hangyujin reviewed (approved):
LGTM

chunhtai commented:
@ValentinVignal looks like the ci is still not happy

ValentinVignal commented:
@chunhtai It should be good now
